tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts

```diff
@@ -646,6 +647,173 @@ describe('metrics store utils', () => {
+        sample: 111,
+      },
+    };
+    let timeSeriesData: TimeSeriesData = {
+      scalars: {},
+      histograms: {},
+      images: {},
+    };
+
+    beforeEach(() => {
+      timeSeriesData = {
+        scalars: {},
+        histograms: {},
+        images: {
+          tagC: {
+            111: {
+              runToLoadState: {},
+              runToSeries: {
+                'test run Id': [
+                  {step: 10, wallTime: 0, imageId: '1'},
+                  {step: 20, wallTime: 10, imageId: '2'},
+                  {step: 30, wallTime: 15, imageId: '3'},
+                ],
+              },
+            },
+          },
+        },
+      };
+    });
+
+    it(`updates cardStepIndex to matched selected time`, () => {
+      const selectedTime = {start: {step: 20}, end: null};
+      const nextCardStepIndex = generateNextCardStepIndexFromSelectedTime(
+        previousCardStepIndex,
+        cardMetadataMap,
+        timeSeriesData,
+        selectedTime
+      );
+
+      expect(nextCardStepIndex).toEqual({[imageCardId]: 1});
+    });
+
+    it(`does not update cardStepIndex on other non-image plugin`, () => {
+      const histogramCardId = 'test histogram card id "plugin":"histogram"';
+      const previousCardStepIndexWtihHistogram = {
+        [histogramCardId]: null,
+      };
+      const cardMetadataMapWtihHistogram = {
+        [histogramCardId]: {
+          runId: 'test run Id',
+          plugin: PluginType.HISTOGRAMS,
+          tag: 'tagB',
+          sample: 111,
+        },
+      };
+      const timeSeriesDataWtihHistogram = {
+        scalars: {},
+        histograms: {
+          tagB: {
+            runToLoadState: {},
+            runToSeries: {
+              'test run Id': [
+                {step: 10, wallTime: 0, bins: [{min: 0, max: 10, count: 2}]},
+                {step: 20, wallTime: 10, bins: [{min: 0, max: 10, count: 2}]},
+                {step: 30, wallTime: 15, bins: [{min: 0, max: 10, count: 2}]},
+              ],
+            },
+          },
+        },
+        images: {},
+      };
+      const selectedTime = {start: {step: 20}, end: null};
+
+      const nextCardStepIndex = generateNextCardStepIndexFromSelectedTime(
+        previousCardStepIndexWtihHistogram,
+        cardMetadataMapWtihHistogram,
+        timeSeriesDataWtihHistogram,
+        selectedTime
+      );
+      expect(nextCardStepIndex).toEqual({
+        [histogramCardId]: null,
+      });
+    });
+
+    it(`does not update cardStepIndex on selected step with no image`, () => {
+      const selectedTime = {start: {step: 15}, end: null};
+      const nextCardStepIndex = generateNextCardStepIndexFromSelectedTime(
+        previousCardStepIndex,
+        cardMetadataMap,
+        timeSeriesData,
+        selectedTime
+      );
+
+      expect(nextCardStepIndex).toEqual({[imageCardId]: null});
+    });
+    it('updates cardStepIndex to smaller closest stepIndexIndex when they are close enough', () => {
```

Comment:
I think this test is redundant - doesn't really add any existing coverage.

Reply:
> But this one tests the combination of getSelectedSteps and getNextImageCardStepIndexFromSingleSelection and also test the logic selectedTime.end === null.

Is that combination already covered by both `updates cardStepIndex to matched selected time` and `does not update cardStepIndex on selected step with no image`? Together those also test the conditions when getNextImageCardStepIndexFromSingleSelection returns non-null and returns null, respectively.

Just to be clear, I don't think you should mock. I'm saying that I believe your first three tests for 'generateNextCardStepIndexFromSelectedTime' cover all the code and interactions in that function (well, except for when `selectedTime.end !== null`, but I understand that will come later). The remaining tests in this group don't test anything that isn't covered elsewhere and are more complex than their alternatives, so I would recommend removing them.